components/calendar/src/types.rs

```diff
@@ -101,56 +83,21 @@ pub enum YearAmbiguity {
     EraAndCenturyRequired,
 }
 
-/// Information about the era as usable for formatting
-///
-/// This is optimized for storing datetime formatting data.
-#[derive(Copy, Clone, Debug, PartialEq)]
-#[non_exhaustive]
-pub enum FormattingEra {
-    /// An Era Index, for calendars with a small, fixed set of eras. The eras are indexed chronologically.
-    ///
-    /// In this context, chronological ordering of eras is obtained by ordering by their start date (or in the case of
-    /// negative eras, their end date) first, and for eras sharing a date, put the negative one first. For example,
-    /// bce < ce, and mundi < incar for Ethiopian.
-    ///
-    /// The TInyStr16 is a fallback string for the era when a display name is not available. It need not be an era code, it should
-    /// be something sensible (or empty).
-    Index(u8, TinyStr16),
-    /// An era code, for calendars with a large set of era codes (Japanese)
-    ///
-    /// This code may not be the canonical era code, but will typically be a valid era alias
-    Code(Era),
-}
-
-impl FormattingEra {
-    /// Get a fallback era name suitable for display to the user when the real era name is not availabe
-    pub fn fallback_name(self) -> TinyStr16 {
-        match self {
-            Self::Index(_idx, fallback) => fallback,
-            Self::Code(era) => era.0,
-        }
-    }
-}
-
 /// Year information for a year that is specified with an era
 #[derive(Copy, Clone, Debug, PartialEq)]
 #[non_exhaustive]
 pub struct EraYear {
-    /// The era code as used in formatting. This era code is not necessarily unique for the calendar, and
-    /// is whatever ICU4X datetime datagen uses for this era.
-    ///
-    /// It will typically be a valid era alias.
-    ///
-    /// <https://tc39.es/proposal-intl-era-monthcode/#table-eras>
-    pub formatting_era: FormattingEra,
-    /// The era code as expected by Temporal/CLDR. This era code is unique for the calendar
-    /// and follows a particular scheme.
-    ///
-    /// <https://tc39.es/proposal-intl-era-monthcode/#table-eras>
-    pub standard_era: Era,
     /// The numeric year in that era
     pub year: i32,
-    /// The ambiguity when formatting this year
+    /// The era code as defined by CLDR, expect for cases where CLDR does not define a code.
+    pub era: TinyStr16,
+    /// An Era Index, for calendars with a small, fixed set of eras. The eras are indexed chronologically.
+    ///
+    /// In this context, chronological ordering of eras is obtained by ordering by their start date (or in the case of
+    /// negative eras, their end date) first, and for eras sharing a date, put the negative one first. For example,
+    /// bce < ce.
+    pub era_index: Option<u8>,
```

Comment:
Nit: Make it super clear that the era index is NOT stable.

Consider naming it something more scary like `formatting_era_index`?

For example, we may be changing it next release for the Hijri and Coptic eras.

Reply:
But it's not formatting-only, it's the index of the era after you sort all eras.

Reply:
It seems risky that there would be drift here. I fully expect that Islamic and Coptic formatting will break after we change the index of an existing era.

Why do we have all this language about chronological ordering? I feel like we should make index 0 be the extended year (primary) era and index 1 be the alternate (secondary) era. There is no calendar with more than 2 eras except Japanese.

Reply:
Idea: change the `Option<u8>` to `Option<FormattingEraKind>` with

```rust
pub enum FormattingEraKind {
    Primary,
    Secondary,
}
```

Reply:
primary era is not a concept that exists yeat

Reply:
Era index is also an ICU4X concept. The concept coincidentally happens to mostly correspond to the separate but strikingly similar concept in UTS 35 called era type, but the two are not derived from each other.

Reply:
The reason I like something more similar to `FormattingEraKind` is that every calendar always has a primary era, and a secondary era is optional. In this way, we can look up the primary era name at index 0 and the secondary era name at index 1.

Reply:
But it's not. Sorting all the eras chronologically to interpret the index can be done by with just CLDR data. Figuring out what the "primary era" is requires ICU4X specific data.

Reply:
What you're proposing is a data model change that can be discussed, but not in this PR. This PR is just moving around fields that already exist while not changing behaviour.

I would prefer this field to be

```rust
/// For calendars with just a primary/secondary era, whether this era is the primary one
is_primary: Option<bool>
```

I wouldn't want to do this before "primary" is a concept in CLDR.

Reply:
(My CLDR PR adds primary era annotations.)

The indexes are also basically hard-coded with ICU4X specific data. Actually it's quite dangerous and broken that icu_calendar hardcodes these indexes when they actually should be data-driven.

Reply:
That's what I added the eras test for, it also checks that the indices match.

Reply:
yeah it's an orthogonal concern. #6444